### ISSUE TITLE
POST json corruption when used as proxy

### ISSUE THREAD

**Issue by masbug** [litespeedtech/openlitespeed#192]
Hi,

I've stumbled upon a bug when using openlitespeed as a proxy. It seems to replace the last character in post data with an X.
Steps to reproduce:
1. Openlitespeed version 1.6.14
2. Setup a (http) web proxy to another server.
3.  ```curl -vvv --http1.1 -i -H "Accept: application/json" -H "Content-Type: application/json; charset=utf-8" --data '{"test":[12.0,48.0]}' http://server/test.php```

test.php:
```
$postdata = file_get_contents("php://input");
print_r(['POST' => $postdata]);
```

Output:
```
Array
(
    [POST] => {"test":[12.0,48.0]X
)
```

Kind regards

**Comment by masbug** [comment]
Hi,

any news on this?
This might be related: https://forum.openlitespeed.org/threads/reverse-proxy-encoding-troubles.4418/

Kind regards

**Comment by raldnor** [comment]
I think it is related, I would like to know an answer as well. I noticed strange behaviour with different services behind a reverse proxy, and it's only when JSON is involved.

**Comment by litespeedtech** [comment]
I tested with 1.7.4 with your php file.
Seems good.
```
user@daviddev:/tmp$ curl -vvv --http1.1 -i -H "Accept: application/json" -H "Content-Type: application/json; charset=utf-8" --data '{"test":[12.0,48.0]}' http://127.0.0.1:8888/test.php
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 8888 (#0)
> POST /test.php HTTP/1.1
> Host: 127.0.0.1:8888
> User-Agent: curl/7.58.0
> Accept: application/json
> Content-Type: application/json; charset=utf-8
> Content-Length: 20
> 
* upload completely sent off: 20 out of 20 bytes
< HTTP/1.1 200 OK
HTTP/1.1 200 OK
< x-powered-by: PHP/5.6.36
x-powered-by: PHP/5.6.36
< content-type: text/html; charset=UTF-8
content-type: text/html; charset=UTF-8
< vary: Accept-Encoding
vary: Accept-Encoding
< transfer-encoding: chunked
transfer-encoding: chunked
< date: Fri, 28 Aug 2020 15:17:07 GMT
date: Fri, 28 Aug 2020 15:17:07 GMT
< server: LiteSpeed
server: LiteSpeed
< connection: Keep-Alive
connection: Keep-Alive

< 
Array
(
    [POST] => {"test":[12.0,48.0]}
)
* Connection #0 to host 127.0.0.1 left intact
```

This is the proxy Vhost setting.

```
docRoot                   $VH_ROOT
  
extprocessor testproxy {
  type                    proxy
  address                 http://10.10.4.32:8088

 maxConns                1000
  initTimeout             10
  retryTimeout            15
  respBuffer              1
}

context / {
  type                    proxy
  handler                 testproxy
  addDefaultCharset       off
}

rewrite  {
  enable                  1
RewriteRule ^/(.*)$ http://testproxy/$1 [P,E=Proxy-Host:10.10.4.32]
}
```

Will test 1.6.16 next.

**Comment by litespeedtech** [comment]
I tested 1.6.16, also good.

**Comment by litespeedtech** [comment]
I tested 1.6.15, same result.
```
user@daviddev:~/ls170/bin$ ./openlitespeed -v
LiteSpeed/1.6.15 Open
	module versions:
	modgzip 1.1
	cache 1.62
	modinspector 1.1
	uploadprogress 1.1
	mod_security 1.3
 (built: Wed Aug  5 17:58:31 UTC 2020) 
	module versions:
	modgzip 1.1
	cache 1.62
	modinspector 1.1
	uploadprogress 1.1
	mod_security 1.3

```
```
user@daviddev:/tmp$ curl -vvv --http1.1 -i -H "Accept: application/json" -H "Content-Type: application/json; charset=utf-8" --data '{"test":[12.0,48.0]}' http://127.0.0.1:8888/test.php
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 8888 (#0)
> POST /test.php HTTP/1.1
> Host: 127.0.0.1:8888
> User-Agent: curl/7.58.0
> Accept: application/json
> Content-Type: application/json; charset=utf-8
> Content-Length: 20
> 
* upload completely sent off: 20 out of 20 bytes
< HTTP/1.1 200 OK
HTTP/1.1 200 OK
< X-Powered-By: PHP/5.6.36
X-Powered-By: PHP/5.6.36
< Content-Type: text/html; charset=UTF-8
Content-Type: text/html; charset=UTF-8
< Vary: Accept-Encoding
Vary: Accept-Encoding
< Transfer-Encoding: chunked
Transfer-Encoding: chunked
< Date: Fri, 28 Aug 2020 16:45:48 GMT
Date: Fri, 28 Aug 2020 16:45:48 GMT
< Server: LiteSpeed
Server: LiteSpeed
< Connection: Keep-Alive
Connection: Keep-Alive

< 
Array
(
    [POST] => {"test":[12.0,48.0]}
)
* Connection #0 to host 127.0.0.1 left intact
```

**Comment by litespeedtech** [comment]
Can you create a ticket so that I can access your server to have a look?

**Comment by masbug** [comment]
I found the difference. If I enable the rewrite, like you did, the error disappears.
`rewrite  {
  enable                  1
  RewriteRule ^/(.*)$ http://testproxy/$1 [P,E=Proxy-Host:10.10.4.32]
}`

If you disable the rewrite then you will be able to reproduce the error.

**Comment by litespeedtech** [comment]
But for a proxy setting, it is supposed to have the rewrite rule set.
I will disable it and do test again and update you later.

**Comment by litespeedtech** [comment]
My test is 1.7.5. Will change to 1.6.15 and test again.

**Comment by litespeedtech** [comment]
I tested 1.6.15 with the same conf, same result.

**Comment by litespeedtech** [comment]
You may create a ticket so that I can access your server to have a look if you want.

**Comment by masbug** [comment]
Thank you for doing all these tests.
I compared your config and mine. I add extra headers in the proxy config. If I remove that then the bug disappears.

Here's my config
```
docRoot                   $VH_ROOT/html/
vhDomain                  test.domain.com
cgroups                   0

errorlog $SERVER_ROOT/logs/$VH_NAME.error.log {
  useServer               0
  logLevel                ERROR
  rollingSize             5M
}

accesslog $SERVER_ROOT/logs/$VH_NAME.access.log {
  useServer               0
  rollingSize             10M
  keepDays                30
  compressArchive         1
}

index  {
  useServer               0
  autoIndex               0
}

accessControl  {
  allow                   *
}

extprocessor test {
  type                    proxy
  address                 192.168.100.100:80
  maxConns                100
  pcKeepAliveTimeout      300
  initTimeout             3
  retryTimeout            2
  respBuffer              0
}

context / {
  type                    proxy
  handler                 test
  extraHeaders            <<<END_extraHeaders
RequestHeader set "X-Forwarded-Ssl" "on"
RequestHeader set "X-Forwarded-Port" "443"
  END_extraHeaders

  addDefaultCharset       off
}

rewrite  {
  enable                  0
  autoLoadHtaccess        0
  logLevel                0
RewriteRule ^/(.*)$ http://test/$1 [P,E=Proxy-Host:192.168.100.100]
}

vhssl  {
  keyFile                 /etc/letsencrypt/live/test.domain.com/privkey.pem
  certFile                /etc/letsencrypt/live/test.domain.com/fullchain.pem
  certChain               1
  sslProtocol             30
  ciphers                 HIGH:!aNULL:!MD5
  enableSpdy              4
  enableStapling          1
}
```



> But for a proxy setting, it is supposed to have the rewrite rule set.

I didn't know. I was following method 1 from [https://openlitespeed.org/kb/reverse-proxy-basics/](https://openlitespeed.org/kb/reverse-proxy-basics/)

**Comment by litespeedtech** [comment]
Sorry for the late response. 
This should have been fixed in 1.6.20 and 1.7.9 release. you can give it a try
`/usr/local/lsws/admin/misc/lsup.sh -b -v 1.6.20`

**Comment by litespeedtech** [comment]
It has been fixed in 1.6.20 and 1.7.9 release.
